zp-relayer/direct-deposit/watcher.ts

```diff
@@ -0,0 +1,98 @@
+// Reference implementation:
+// https://github.com/omni/tokenbridge/blob/master/oracle/src/watcher.js
+import type Web3 from 'web3'
+import type { AbiItem } from 'web3-utils'
+import { web3 } from '@/services/web3'
+import PoolAbi from '@/abi/pool-abi.json'
+import config from '@/configs/watcherConfig'
+import { logger } from '@/services/appLogger'
+import { redis } from '@/services/redisClient'
+import { DirectDeposit, poolTxQueue, WorkerTxType, WorkerTxTypePriority } from '@/queue/poolTxQueue'
+import { lastProcessedBlock, getLastProcessedBlock, updateLastProcessedBlock, parseDirectDepositEvent } from './utils'
+import { BatchCache } from './BatchCache'
+import { validateDirectDeposit } from '@/validation/tx/validateDirectDeposit'
+import { getBlockNumber, getEvents } from '@/utils/web3'
+
+const PoolInstance = new web3.eth.Contract(PoolAbi as AbiItem[], config.poolAddress)
+
+const eventName = 'SubmitDirectDeposit'
+
+const batch = new BatchCache<DirectDeposit>(
+  config.directDepositBatchSize,
+  config.directDepositBatchTtl,
+  ds => {
+    logger.info('Adding direct-deposit events to queue', { count: ds.length })
+    poolTxQueue.add(
+      '',
+      {
+        transactions: [
+          {
+            deposits: ds,
+          },
+        ],
+        type: WorkerTxType.DirectDeposit,
+        // TODO: traceId
+      },
+      {
+        priority: WorkerTxTypePriority[WorkerTxType.DirectDeposit],
+      }
+    )
+  },
+  dd => validateDirectDeposit(dd, PoolInstance),
+  redis
+)
+
+async function init() {
+  await getLastProcessedBlock()
+  await batch.init()
+  runWatcher()
+}
+
+async function getLastBlockToProcess(web3: Web3) {
+  const lastBlockNumber = await getBlockNumber(web3)
+  return lastBlockNumber
+}
+
+async function watch() {
+  const lastBlockToProcess = await getLastBlockToProcess(web3)
+
+  if (lastBlockToProcess <= lastProcessedBlock) {
+    logger.debug('All blocks already processed')
+    return
+  }
+
+  const fromBlock = lastProcessedBlock + 1
+  const rangeEndBlock = fromBlock + config.eventsProcessingBatchSize
+  let toBlock = Math.min(lastBlockToProcess, rangeEndBlock)
+
+  let events = await getEvents(PoolInstance, eventName, {
```

Comment:
Could you elaborate? Do you mean the case when watcher misses some event and continues processing?

Reply:
Since we are not using "required block confirmation" in the current code, even a small reorg can lead to event loss. Maybe we can return "required block confirmation" back, and then think about detecting and recovering from gaps in nonces under a separate PR in the future? wdyt?

Reply:
>Maybe we can return "required block confirmation" back

Totally agree. Will prepare changes.

>think about detecting and recovering from gaps in nonces

Still, don't get how we can get a "gap in event  nonces"?

Reply:
Added block confirmations for watcher

Reply:
1. Imagine we had requested all events up to block 100, which included event with nonce 5.
2. Events with nonces 6 and 7 happened in block 101.
3. Then, a short reorg happened, and transaction with event 6 moved from block 101 to block 100.
4. But as we will only request events from block 101 during the next cycle, we won't see it. Instead we will only observe event with nonce 7.
5. Since the nonces are sequential, I assume we can create a reliable mechanism for recovering from such issues.

Reply:
Got it. So, we probably need to track nonces, as well, to prevent such cases. I think this indeed should me moved too a separate issue addressed later.